Auto-authorize Jetpack connections for Woo users

v18 opened the pull request:
This PR sets `autoAuthenticate` to `true` for users coming from the WooCommerce Services flow that includes privacy information about Jetpack.

Here's a video showing what it looks like:

https://cloudup.com/cpZ-fOFL7yV

Background info: p7bbVw-1Rl-p2

I tested the flow using a logged out and logged in user, and I tested the usual Jetpack flow as logged in + out, and from wordpress.com/jetpack/connect

# Testing Instructions
1. Install Jetpack on your site, and deactivate it
2. In the file `class.jetpack.php`, find the function `build_connect_url` and add this line to it right above `return $raw ? $url : esc_url( $url );`:  `$url = add_query_arg( 'calypso_env', 'development', $url );`
3. Check out the branch `update/wcs-jp-flow` from https://github.com/Automattic/woocommerce-services and run `npm start` - https://github.com/Automattic/woocommerce-services/pull/1114
4. Go to the plugins page on your site, and click the WooCommerce Services connect button to see the new flow

tyxla commented on client/state/jetpack-connect/reducer.js:
```diff
@@ -182,10 +188,12 @@ export function jetpackConnectAuthorize( state = {}, action ) {
 				buildDefaultAuthorizeState(),
-				{ queryObject: queryObject }
+				{ queryObject: queryObject },
+				{ autoAuthorize: shouldAutoAuthorize }
```
Any reason to have `autoAuthorize: shouldAutoAuthorize` in a separate, new object, rather than in the previous one that we're assigning from?

tyxla replied:
Also, another concern here is that we're overwriting `autoAuthorize` in all cases. Wouldn't this unexpectedly overwrite the existing value in case? For example in the case when creating an account in the JPC flow (auto authorization is supposed to be enabled in that case, but it can get inadvertently disabled when we're not coming from the WooCommerce Services flow). Do we want to be more careful here and don't touch auto authorization if we're outside of the WooCommerce Services flow?

v18 commented:
Thanks for the comments @tyxla! You mentioned using `queryObject` instead of a new `autoAuthorizing`. I'm not sure if this is what you had in mind, but I thought of doing this instead:

In `<LoggedInForm />` (https://github.com/Automattic/wp-calypso/blob/master/client/jetpack-connect/auth-logged-in-form.jsx#L80), we can check for `'woocommerce-services-auto-authorize' === queryObject.from` when checking whether to autoAuthorize or not.

This will not affect any other flow - as we'll still be checking for all the previously-used props and not overriding any of them.

What do you think?

tyxla commented:
Sorry @v18, I might have been not clear enough. Let me try to clarify.

The `JETPACK_CONNECT_QUERY_SET` action is called in all of our flows that include authorization.    In your case, `const shouldAutoAuthorize = ( 'woocommerce-services-auto-authorize' === get( queryObject, 'from' ) );` will be `true`, but in the rest of the cases it will be `false`. And if you always override `autoAuthorize` with this value, you're risking to alter the `autoAuthorize` that might have been set to `true` from a previously called action, called before setting the query args.

Not sure if that clarifies it well, so let me know if further details are needed 😉

v18 commented:
> And if you always override autoAuthorize with this value, you're risking to alter the autoAuthorize that might have been set to true from a previously called action

@tyxla I started to look at this again, and I wrote a few unit tests to make sure that `autoAuthorize` does not get overriden from a previous value. The unit test for not overriding it, though, fails using the original code. So it seems like when `JETPACK_CONNECT_QUERY_SET` is processed, `state.autoAuthorize` is always reset to `undefined`.

I removed the previous code (for easier unit testin') in 40d5bc5538ac0e761cab270047917521b20d5bbb

My unit tests are in 5116d524b7db194f7c22f0bc521a99da7c31f93d

The specific one that checks that a `true` autoAuthorize is not overriden to false is here: 

https://github.com/Automattic/wp-calypso/pull/16937/files#diff-af4ef7de2620b88a2c715c5a6a860edcR661

I also added a new commit that makes the unit tests pass, and follows your suggestion not to override autoAuthorize: 

3194b33e1e4aa8f52647d78aa1ca32f52dc00b71

I wonder now if the previous behaviour of overriding `autoAuthorize` was needed for something, or if we're correct in preserving `state.autoAuthorize` when handling `JETPACK_CONNECT_QUERY_SET`. What do you think?

tyxla commented:
Generally, this looks like a good approach to me 👍  I have another tip, sorry if it's confusing.

> I wonder now if the previous behaviour of overriding autoAuthorize was needed for something

Looking further in the cases, it seems it's not needed, because it gets overridden after that in of the cases when we need to auto authorize. It would probably be a good solution to just add a `autoAuthorize: false` to `buildDefaultAuthorizeState()` - let's also ask @johnHackworth if this makes sense.

If we go that way, I'd recommend:

* Not passing `state.autoAuthorize` anymore, as it might not be accurate - the defaults should come from `buildDefaultAuthorizeState()`.
* Building an object depending on the query, and assigning it to the state, like this:

```
const additionalQueryArgs = 'woocommerce-services-auto-authorize' === get( queryObject, 'from' )
	? { autoAuthorize: true }
	: {};

return Object.assign(
	{},
	buildDefaultAuthorizeState(),
	{ queryObject: queryObject },
	additionalQueryArgs,
);
```

> or if we're correct in preserving state.autoAuthorize when handling `JETPACK_CONNECT_QUERY_SET`

because this will surely leave `autoAuthorize` intact if you don't really intend to modify it.

In practice, `autoAuthorize` will be overridden later if it needs to be `true`, so there isn't much trouble around it. Trouble might come if we eventually decide to use `JETPACK_CONNECT_QUERY_SET` for another use case (intercepting parameters from somewhere else but the URL query args) - but we might never want to do that, so IMHO it's not a concern.

v18 commented:
> It would probably be a good solution to just add a autoAuthorize: false to buildDefaultAuthorizeState() - let's also ask @johnHackworth if this makes sense.

@johnHackworth do you have an opinion on this?

> Building an object depending on the query, and assigning it to the state

Cool, can do based on the decision we make.


> Trouble might come if we eventually decide to use JETPACK_CONNECT_QUERY_SET for another use case (intercepting parameters from somewhere else but the URL query args) - but we might never want to do that, so IMHO it's not a concern.

Yeh, I agree that we shouldn't solve this potential problem now 👍 . If things change that affect this, then it should be handled in that case.

v18 commented:
@apeatling I don't think this PR needs to be merged anymore, can close.

v18 commented:
Hi @tyxla, I made these changes:

1. Added `autoAuthorize: false` to `buildDefaultAuthorizeState()`
2. Built an additional query object in the case `JETPACK_CONNECT_QUERY_SET`
3. Updated the tests
4. Added a new parameter we're using that's eligible for auto-authorization
4. Rebased on master

This covers all of your concerns so far. I tested using these scenarios:
1. Normal Jetpack connection - does not auto authorize
2. The Jetpack connection from the WooCommerce wizard at `/wp-admin/index.php?page=wc-setup&step=activate` (this is currently live and working)

Let me know if this is good to go. Thanks!

tyxla commented on client/state/jetpack-connect/reducer/jetpack-connect-authorize.js:
```diff
@@ -75,7 +76,17 @@ export default function jetpackConnectAuthorize( state = {}, action ) {
+			return Object.assign(
+				{},
+				buildDefaultAuthorizeState(),
+				{ queryObject: queryObject },
```
We can shorten `{ queryObject: queryObject }` to `{ queryObject }`

v18 replied:
👍 Fixed & rebased.

seear commented:
Looks great to me.

One thing I noticed is that `buildDefaultAuthorizeState()` is actually only used in one place, so it might make sense to get rid of it and set everything explicitly in the `JETPACK_CONNECT_QUERY_SET` case (setting `autoAuthorize` appropriately).

I quickly tested the auto-auth case of connecting a new jetpack site from Calypso and it worked fine.

v18 commented:
@seear thought I see your point, I'd prefer not to change `buildDefaultAuthorizeState()` in this PR so I can quickly merge it.
